src/firmware/app/primitives/direct_control_primitive.c

```diff
@@ -0,0 +1,147 @@
+#include "firmware/app/primitives/direct_control_primitive.h"
+
+#include <assert.h>
+
+#include "firmware/app/control/control.h"
+
+typedef struct DirectControlPrimitiveState
+{
+    bool direct_velocity;
+    float direct_target_velocity_x;
+    float direct_target_velocity_y;
+    float direct_target_velocity_angular;
+} DirectControlPrimitiveState_t;
+DEFINE_PRIMITIVE_STATE_CREATE_AND_DESTROY_FUNCTIONS(DirectControlPrimitiveState_t)
+
+void app_direct_control_primitive_start(DirectControlPrimitiveMsg prim_msg,
+                                        void* void_state_ptr, FirmwareWorld_t* world)
+{
+    DirectControlPrimitiveState_t* state = (DirectControlPrimitiveState_t*)void_state_ptr;
+    const FirmwareRobot_t* robot         = app_firmware_world_getRobot(world);
+    Chicker_t* chicker                   = app_firmware_robot_getChicker(robot);
+    Dribbler_t* dribbler                 = app_firmware_robot_getDribbler(robot);
+    Charger_t* charger                   = app_firmware_robot_getCharger(robot);
+
+    switch (prim_msg.which_wheel_control)
+    {
+        case DirectControlPrimitiveMsg_direct_per_wheel_control_tag:
+        {
+            DirectControlPrimitiveMsg_DirectPerWheelControlMsg control_msg =
+                prim_msg.wheel_control.direct_per_wheel_control;
+            state->direct_velocity = false;
+            // TODO (#1649): Fix passing rpm into an applyForce function
+            app_wheel_applyForce(app_firmware_robot_getFrontLeftWheel(robot),
+                                 control_msg.front_left_wheel_rpm);
+            app_wheel_applyForce(app_firmware_robot_getBackLeftWheel(robot),
+                                 control_msg.back_left_wheel_rpm);
+            app_wheel_applyForce(app_firmware_robot_getBackRightWheel(robot),
+                                 control_msg.back_right_wheel_rpm);
+            app_wheel_applyForce(app_firmware_robot_getFrontRightWheel(robot),
+                                 control_msg.front_right_wheel_rpm);
+            break;
+        }
+        case DirectControlPrimitiveMsg_direct_velocity_control_tag:
+        {
+            DirectControlPrimitiveMsg_DirectVelocityControlMsg control_msg =
+                prim_msg.wheel_control.direct_velocity_control;
+            state->direct_velocity          = true;
+            state->direct_target_velocity_x = control_msg.velocity.x_component_meters;
+            state->direct_target_velocity_y = control_msg.velocity.y_component_meters;
+            state->direct_target_velocity_angular =
+                control_msg.angular_velocity.radians_per_second;
+            break;
+        }
+        default:
+        {
+            // Do nothing
+            state->direct_velocity = false;
+            assert(false);
```

Comment:
Do we want this to abort or to actually do nothing?

Reply:
hm, could just log an error here...

Reply:
That seems like the best option. Some sort of generic logging mechanism would be good, but definitely out-of-scope here. Maybe we already have a ticket for one? If not this might be a good time to add one.

Reply:
added TODO with https://github.com/UBC-Thunderbots/Software/issues/1652